### PR TITLE
Peer storage feature

### PR DESCRIPTION
**Pull request by adi2011**
# PEER STORAGE BACKUP

This PR implements `peer storage backup` which will enable nodes to exchange their respective SCBs (Static channel backup), This will be useful in case of complete data loss.

### WHY

One of the major features of lightning is that the transactions are off-chain and are stored in a local database, which makes this DB highly dynamic and complex to backup.
Peer storage backup will allow users to send their encrypted backup to their peers, which can be used to recover their funds in case of complete data loss.

### HOW

The strategy is that we'll store the data received in the [datastore](https://lightning.readthedocs.io/lightning-datastore.7.html).
So to implement this I've introduced 2 new messages i.e. `PEER_STORAGE` and `YOUR_PEER_STORAGE`, these will be used to exchange the data stored between the nodes. 
-`PEER_STORAGE` will be used to send our own encrypted backup to the peer.
-`YOUR_PEER_STORAGE` will be used to send the most recent backup of the peer.

The general flow of messages every time we `connect` is:

Alice -----------------------  :hearts: ------------------------------ Bob
-------------------------------------------------------------------------------------------------------
PEER_STORAGE :outbox_tray:-------------:incoming_envelope:-------------> :inbox_tray:
YOUR_PEER_STORAGE :outbox_tray:---------:incoming_envelope:------------> :inbox_tray:

:inbox_tray: <------------:incoming_envelope:--------------PEER_STORAGE :outbox_tray:
:inbox_tray: <-----------:incoming_envelope:------------YOUR_PEER_STORAGE :outbox_tray: 

-------------------------------------------------------------------------------------------------------

On receiving `YOUR_PEER_STORAGE` bob will verify if it's correct and Alice has not changed anything in his last sent backup.
On receiving `PEER_STORAGE` bob will update the backup he has stored for Alice in his own datastore.

Every time we open a new channel or close an old one we will send `PEER_STORAGE` to every peer we're connected to (Haven't figured out yet, maybe a new RPC (`sendcustommsgmulti`) which will enable plugins to send a single message to multiple users at once, because the interaction between lightningd and plugins is single-threaded)

They can choose to ignore the messages since they are odd (`It's okay to be odd`)

In case of complete data loss, the user will reconnect to their peers and hope that they get a YOUR_PEER_STORAGE message. Then they can directly use the RPC specified in the plugin to recover the channels.

### CODE REVIEW

File: tests/test_misc.py
Comment (vincenzopalazzo) [inline tests/test_misc.py:2277]:
```suggestion
    secret = l1.rpc.makesecret("To the moon!")["secret"]
```

cit: dogecoin 😄

File: lightningd/peer_control.c
Comment (vincenzopalazzo) [inline lightningd/peer_control.c:1765]:
```suggestion
              return command_param_failed();
```

missed a tab here!

File: lightningd/peer_control.c
Comment (vincenzopalazzo) [inline lightningd/peer_control.c:1784]:
```suggestion
	json_staticbackup,
```

There is any reason to do not this? maybe I'm missing some previous discussion, just asking because I use a lot `git grep "json_{command_name}"` to search some command implementation

**Review by vincenzopalazzo — COMMENTED** [review]
Concept ack I just read the code, it requires more love in the review from my side!

**Comment by adi2011** [comment]
Thanks for the review @vincenzopalazzo, I have mostly fixed all the mentioned changes in the SCB PR, This one will get in after that :)

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:786]: Could be more descriptive of what this does?

```
Checks if i have got a backup from a peer, and if so, will stub those channels in the database and if is successful, will return list of channels that have been successfully stubbed
```

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:724]: nit: call it `scb` not `bkp`

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:709]: nit: rename `bkp` to `scb`

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:669]: `no SCB`

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:690]: who's the `admin`? 

`incompatible scb file version on disk`.

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:685]: *on disk.

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:565]: can we add a comment about the two types of custom messages that are valid here?

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:571]: you can just pull a node_id? 

also check out json_scan

```
err = json_scan(tmpctx, buf, params,
                           "{peer_id:%, payload:%}",
                           JSON_SCAN(json_to_node_id, &node_id),
                           JSON_SCAN_TAL(cmd, json_strdup, hexdata));
                           
 if (err)
       error out somehow
 
payload = tal_hex(cmd, hexdata);
```

File: plugins/chanbackup.c
Comment (niftynei) [inline plugins/chanbackup.c:482]: This should def be in it's own commit.

**Review by niftynei — COMMENTED** [review]
Half done review, pushing up comments.

In general, smaller commits for each independent change would be great.

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:784]: `restorefrompeer` looks good :+1:, `-` is not allowed I think, they create problems in other languages.

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:690]: > who's the `admin`?
> you

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:565]: Yes, I'll add comments and write a doc as well for this

**Review by niftynei — COMMENTED** [review]
some comments!

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:811]: Okay, these are some really nice questions:
Ans. 1) This would be used in case of data loss. The user could connect to their peers and collect their storage.

Ans. 2) Hmm, That's an excellent suggestion, everytime we receive scb we can automatically stub the channels in the DB (which are not already into it).

Ans. 3) This feature would be used as a last resort to recover the funds in case of severe data loss, This can only help users to recover funds in case of severe data loss, This would never harm existing channels and funds.

Ans. 4) We skip over them and do not stub them in the DB.

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:715]: If they have an older version, they can contact the team to see if it could be parsed and used for fund recovery.

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c]: > filenames in line is not great either fwiw.

Defined a global var for it.

Comment (adi2011) [reply plugins/chanbackup.c]: > this really could be a separate function? don't you do this other places?

Hmm, Done!

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:835]: Done

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c]: Removed, Was using it for debugging

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c]: Done, we just need the index. Libplugin automatically handles multiple requests one by one

**Comment by cdecker** [comment]
@adi2011: this is marked for release 22.11 for which we're trying to publish a first RC this week, but it is also marked as a draft. What's the current status of this? Happy to review and merge it if it is ready, otherwise we can also postpone it to the next release, which'll happen in ~2 months again.

**Comment by cdecker** [comment]
After talking to @adi2011 we decided that we'll be pushing this PR to the next release.

**Comment by adi2011** [comment]
Thanks! Will get onto it, once my semester exams are over.

File: plugins/chanbackup.c
Comment (rustyrussell) [inline plugins/chanbackup.c:390]: We want a path here, not just top-level keys in the datrastore!

i.e. ["chanbackup", "peers", nodeid].  In fact, you can see how autoclean does it with "datastore_path", though we do't have datastore helpers for this.  Let me write some...

File: plugins/chanbackup.c
Comment (rustyrussell) [inline plugins/chanbackup.c:745]: OK, so I added more some datastore helpers in https://github.com/ElementsProject/lightning/pull/5809 (we already had ones to set datastore entries, which were included since you created this PR, and you should use those too).

File: plugins/chanbackup.c
Comment (rustyrussell) [inline plugins/chanbackup.c]: I guess this is arguable?  It's not terrible to restart it, for example.  It still works...

File: plugins/chanbackup.c
Comment (rustyrussell) [inline plugins/chanbackup.c:440]: Definitely time to create a "grab_peer_storage(const tal_t *ctx, struct plugin *plugin)" that does all this work (using grab_file, ofc!).

File: connectd/multiplex.c
Comment (adi2011) [reply connectd/multiplex.c:728]: Agreed, done!

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:835]: ```struct feature_set *feature_set_for_feature(const tal_t *ctx, int feature) ``` works for compulsory_features() only, These are odd-numbered features, I don't think it's possible to use this here?

Comment (adi2011) [reply plugins/chanbackup.c:835]: Done

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:524]: When I declare this on the stack, the passed value to `after_send_scb` changes for some reason.

Comment (adi2011) [reply plugins/chanbackup.c:524]: Declaring this on the stack modifies it's value in the `after_send_scb()` because it is passed as a pointer.

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c:577]: Done!

**Comment by adi2011** [comment]
Need datastore helpers to set a path for the backup.

**Comment by adi2011** [comment]
DONE!

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c]: `Custom featurebits only allows for non-dynamic.`

**Review by rustyrussell — APPROVED** [review]
I need to think more about the test failure: because we send an error as soon as the peer send reestablish, we never get to restore our backup!  I think the simplest workaround is to wait a second or so before sending the error in this case, but I'll think more...

**Comment by rustyrussell** [comment]
Also needs to shuffle into fewer commits, but I can go through that with you once we're happy with everything and it's all reviewed.

File: connectd/multiplex.c
Comment (adi2011) [reply connectd/multiplex.c]: Hmm already done in a later commit, will rebase it once it passes the tests...

**Comment by rustyrussell** [comment]
See also: #5903

**Comment by rustyrussell** [comment]
OK, I played with the tests, and added some more fixes.  I've pushed them up here:

https://github.com/rustyrussell/lightning/tree/guilt/pr-5361

**Review by m-schmoock — DISMISSED** [review]
Hey @adi2011 , just some things I noticed...

File: plugins/chanbackup.c
Comment (rustyrussell) [reply plugins/chanbackup.c:178]: Yes, but grab_file() added a NUL term, which is why we remove it here.

**Comment by rustyrussell** [comment]
OK, I shuffled these into a reasonable series of commits, rebased on master, and inverted the peer_wire_is_internal return (it was definitely backwards for my reading!).

The result is here:

https://github.com/rustyrussell/lightning/tree/guilt/pr-5361

If you like that @adi2011 you can simply force push that to your branch!

File: plugins/chanbackup.c
Comment (m-schmoock) [reply plugins/chanbackup.c:176]: Oh, I misread the comment. This is supposed to remove it ofc.

File: plugins/chanbackup.c
Comment (m-schmoock) [reply plugins/chanbackup.c:178]: I misread the comment, ofc this is supposed to remove the nul term.

**Comment by rustyrussell** [comment]
OK, this needs to be behind an experimental config setting (also avoids much test churn!).  Four more cleanups are here:

https://github.com/rustyrussell/lightning/commits/guilt/pr-5361

File: plugins/chanbackup.c
Comment (rustyrussell) [reply plugins/chanbackup.c:262]: Generally, I prefer {} around any multi-line thing, even if it's a single statement.  There are some coding styles that insist on {} in *all* cases, to reduce errors, but I've only ever had a real bug caused by wrong braces once in my entire coding career, so I don't think it's a huge priority.

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/commit/0079895cacb0567ce15bd1f1436ce1f9c1e1e502

**Comment by rustyrussell** [comment]
Fix dumb code typo, and put the fix earlier in the PR so it doesn't break bisect!

**Comment by endothermicdev** [comment]
ACK a54dcfb44ed6f8b03ae918b7425ab9d7fddf644b
Just waiting on CI now.

**Comment by rustyrussell** [comment]
Seems like it got wedged in the source check somehow?  I cannot cancel it either.  Pushing trivial change to kick it.

Ack a4f38b7c1d87a8888483ed14dc2f7a01351b8e96

File: plugins/chanbackup.c
Comment (m-schmoock) [inline plugins/chanbackup.c:174]: I think we should use `FILENAME` here as well instead of the similar string. This way, when filename get changed, code does not break.

**Review by m-schmoock — COMMENTED** [review]
Looking good now

**Comment by rustyrussell** [comment]
Dumb typo completely broke peer_connected hook!

Ack 5805664d79812431a6fd778d710573d1a046c463

File: plugins/chanbackup.c
Comment (adi2011) [reply plugins/chanbackup.c]: Agreed. Thank you for the review!

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/commit/081a1d847ff287c02ebc093e1d524f61d588d118

**Comment by endothermicdev** [comment]
Congrats @adi2011!  It was a long road, but this is a very cool feature!

**Comment by Sjors** [comment]
Is there any documentation for this, other than the name of the option?

In particular it would be useful to illustrate how recovery works, e.g. "Start a new node with the same secret, find one of your old peers e.g. using a public archive like mempool . space, call  lightning-cli restore-from-peer, etc".

`man lightning-restore-from-peer` is not working for me either.

**Pull request merged**
